---
title: "Tutoriel DADA2"
output: github_document
---

### Téléchargement de DADA2 :

```{r}
library(dada2)

```

### Unzip des données :

```{r}
path <- "/home/rstudio/Tuto DADA 2/Tuto-DADA2/MiSeq_SOP"
list.files(path)
```

### Regroupement des forward et reverse :

```{r}
fnFs <- sort(list.files(path, pattern= "_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern= "_R2_001.fastq", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
plotQualityProfile(fnFs[1:2])
```

Graphique représentant les profiles de qualité des séquences forward.

```{r}
plotQualityProfile(fnRs[1:2])
```

Graphique représentant les profiles de qualité des séquences reverse.

### Création d'un nouveau dossier "filtered" dans le directory et filtration des séquences :

```{r }
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240, 160), maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, compress = TRUE, multithread = TRUE)

head(out)
```

### Création du modèle de calcul des erreurs à partir du jeu de données :

```{r}
errF <- learnErrors(filtFs, multithread = TRUE)
errR <- learnErrors(filtRs, multithread = TRUE)

plotErrors(errF, nominalQ = TRUE)
```

### Correction des erreurs et détection des vrais variants :

```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)

dadaFs[[1]]
```

### Alignement des séquences forward et reverse :

```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
head(mergers[[1]])
```

### Construction table d'observation

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
```

```{r}
table(nchar(getSequences(seqtab)))
```

### Elimination des chimères :

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method = "consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
```

```{r}
sum(seqtab.nochim)/sum(seqtab)
```

### Tableau récapitulatif du nombre de séquences enlevées à chaque étape :

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```
### Taxonomie

```{r}
taxa <- assignTaxonomy(seqtab.nochim, "silva_nr99_v138.1_train_set.fa.gz", multithread=TRUE)
taxa.print <- taxa
row.names(taxa.print) <- NULL
head(taxa.print)
```
### Mesure de la précision de DADA2

```{r}
unqs.mock <- seqtab.nochim["Mock",]
unqs.mock <- sort(unqs.mock[unqs.mock>0], decreasing=TRUE)
cat("DADA2 inferred", length(unqs.mock), "sample sequences present in the Mock community.\n")
```

```{r}
mock.ref <- getSequences(file.path(path, "HMP_MOCK.v35.fasta"))
match.ref <- sum(sapply(names(unqs.mock), function(x) any(grepl(x, mock.ref))))
cat("Of those,", sum(match.ref), "were exact matches to the expected reference sequences.\n")
```

